File: MeepMeepTesting/src/main/java/com/example/meepmeeptesting/EditablePose.java
package com.example.meepmeeptesting;

import static com.example.meepmeeptesting.MeepMeepTesting.X_START_LEFT;
import static com.example.meepmeeptesting.MeepMeepTesting.X_START_RIGHT;
import static com.example.meepmeeptesting.MeepMeepTesting.backdropSide;
import static com.example.meepmeeptesting.MeepMeepTesting.isRed;

import com.acmerobotics.roadrunner.geometry.Pose2d;

class EditablePose {

    public double x, y, heading;

    public EditablePose(double x, double y, double heading) {
        this.x = x;
        this.y = y;
        this.heading = heading;
    }

    public EditablePose byAlliance() {
        double alliance = isRed ? 1 : -1;
        y *= alliance;
        heading *= alliance;
        return this;
    }

    public EditablePose bySide() {
        if (!backdropSide) x += X_START_LEFT - X_START_RIGHT;
        return this;
    }

    public EditablePose byBoth() {
        return byAlliance().bySide();
    }

    public Pose2d toPose2d() {
        return new Pose2d(x, y, heading);
    }

    public EditablePose flipBySide() {
        if (!backdropSide) heading = Math.PI - heading;
        if (!backdropSide) x = (X_START_LEFT + X_START_RIGHT) - x;
        return this;
    }
}

File: MeepMeepTesting/src/main/java/com/example/meepmeeptesting/EditablePose.kt
package com.example.meepmeeptesting

import com.acmerobotics.roadrunner.geometry.Pose2d
import com.example.meepmeeptesting.MeepMeepTesting.*

data class EditablePose

@JvmOverloads
constructor(
    @JvmField var x: Double = 0.0,
    @JvmField var y: Double = 0.0,
    @JvmField var heading: Double = 0.0,
) {
    fun byAlliance(): EditablePose {
        val alliance: Double = if (isRed) 1.0 else -1.0
        y *= alliance
        heading *= alliance
        return this
    }

    fun bySide(): EditablePose {
        if (backdropSide != isRed) x += X_START_LEFT - X_START_RIGHT
        return this
    }

    fun byBoth(): EditablePose {
        return byAlliance().bySide()
    }

    fun toPose2d(): Pose2d {
        return Pose2d(x, y, heading)
    }

    fun flipBySide(): EditablePose {
        if (!backdropSide) heading = Math.PI - heading
        if (!backdropSide) x = (X_START_LEFT + X_START_RIGHT) - x
        return this
    }
}